frame/contracts/src/rent.rs

```diff
@@ -14,64 +14,87 @@
 // You should have received a copy of the GNU General Public License
 // along with Substrate. If not, see <http://www.gnu.org/licenses/>.
 
-use crate::{Module, RawEvent, BalanceOf, ContractInfo, ContractInfoOf, TombstoneContractInfo,
-	Trait, AliveContractInfo};
-use sp_runtime::traits::{Bounded, CheckedDiv, CheckedMul, Saturating, Zero,
-	SaturatedConversion};
-use frame_support::traits::{Currency, ExistenceRequirement, Get, WithdrawReason, OnUnbalanced};
-use frame_support::StorageMap;
+use crate::{
+	AliveContractInfo, BalanceOf, ContractInfo, ContractInfoOf, Module, RawEvent,
+	TombstoneContractInfo, Trait,
+};
 use frame_support::storage::child;
+use frame_support::traits::{Currency, ExistenceRequirement, Get, OnUnbalanced, WithdrawReason};
+use frame_support::StorageMap;
+use sp_runtime::traits::{Bounded, CheckedDiv, CheckedMul, SaturatedConversion, Saturating, Zero};
 
-#[derive(PartialEq, Eq, Copy, Clone)]
-#[must_use]
-pub enum RentOutcome {
-	/// Exempted from rent iff:
-	/// * rent is offset completely by the `rent_deposit_offset`,
-	/// * or rent has already been paid for this block number,
-	/// * or account doesn't have a contract,
-	/// * or account has a tombstone.
-	Exempted,
-	/// Evicted iff:
-	/// * rent exceed rent allowance,
-	/// * or can't withdraw the rent,
-	/// * or go below subsistence threshold.
-	Evicted,
-	/// The outstanding dues were paid or were able to be paid.
-	Ok,
+/// The amount to charge.
+///
+/// This amount respects the contract's rent allowance and the subsistence deposit.
+/// Because of that, charging the amount cannot remove the contract.
+struct OutstandingAmount<T: Trait> {
+	amount: BalanceOf<T>,
 }
 
-/// Evict and optionally pay dues (or check account can pay them otherwise) at the current
-/// block number (modulo `handicap`, read on).
-///
-/// `pay_rent` gives an ability to pay or skip paying rent.
-/// `handicap` gives a way to check or pay the rent up to a moment in the past instead
-/// of current block.
+impl<T: Trait> OutstandingAmount<T> {
+	/// Create the new outstanding amount.
+	///
+	/// The amount should be always withdrawable and it should not kill the account.
+	fn new(amount: BalanceOf<T>) -> Self {
+		Self { amount }
+	}
+
+	/// Returns the amount this instance wraps.
+	fn peek(&self) -> BalanceOf<T> {
+		self.amount
+	}
+
+	/// Withdraws the outstanding amount from the given account.
+	fn withdraw(self, account: &T::AccountId) {
+		if let Ok(imbalance) = T::Currency::withdraw(
+			account,
+			self.amount,
+			WithdrawReason::Fee.into(),
+			ExistenceRequirement::KeepAlive,
+		) {
+			// This should never fail. However, let's err on the safe side.
+			T::RentPayment::on_unbalanced(imbalance);
+		}
+	}
+}
+
+enum Verdict<T: Trait> {
+	Exempt,
```

Comment:
This one does not deserve doc comment?

Reply:
Hah, before wrapping up I looked at it and thought that it's pretty clear.

Now, however, I don't think so )

Reply:
Done!